flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBase.java

```diff
@@ -17,130 +17,191 @@
-    private transient PreparedStatement prepared;
-    private transient FutureCallback<ResultSet> callback;
-    private transient Throwable exception = null;
+    protected transient Session session;
+    private transient FutureCallback<V> callback;
+    private AtomicReference<Throwable> throwable;
 
-    public CassandraOutputFormatBase(String insertQuery, ClusterBuilder builder) {
-        Preconditions.checkArgument(
-                !Strings.isNullOrEmpty(insertQuery), "Query cannot be null or empty");
+    public CassandraOutputFormatBase(ClusterBuilder builder) {
         Preconditions.checkNotNull(builder, "Builder cannot be null");
-
-        this.insertQuery = insertQuery;
         this.builder = builder;
     }
 
+    /**
+     * Sets the maximum allowed number of concurrent requests for this output format.
+     *
+     * @param maxConcurrentRequestsTimeout timeout duration when acquiring a permit to execute
+     */
+    public void setMaxConcurrentRequestsTimeout(Duration maxConcurrentRequestsTimeout) {
+        Preconditions.checkNotNull(
+                maxConcurrentRequestsTimeout, "Max concurrent requests timeout cannot be null");
+        Preconditions.checkArgument(
+                !maxConcurrentRequestsTimeout.isNegative(),
+                "Max concurrent requests timeout is expected to be positive");
+
+        this.maxConcurrentRequestsTimeout = maxConcurrentRequestsTimeout;
+    }
+
+    /**
+     * Sets the maximum allowed number of concurrent requests for this output format.
+     *
+     * @param maxConcurrentRequests maximum number of concurrent requests allowed
+     */
+    public void setMaxConcurrentRequests(int maxConcurrentRequests) {
+        Preconditions.checkArgument(
+                maxConcurrentRequests > 0, "Max concurrent requests is expected to be positive");
+        this.maxConcurrentRequests = maxConcurrentRequests;
+    }
+
+    /**
+     * Configure the connection to Cassandra.
+     *
+     * @param parameters The configuration with all parameters.
+     */
     @Override
     public void configure(Configuration parameters) {
         this.cluster = builder.getCluster();
     }
 
     /**
-     * Opens a Session to Cassandra and initializes the prepared statement.
+     * Opens a Session to Cassandra .
      *
      * @param taskNumber The number of the parallel instance.
-     * @throws IOException Thrown, if the output could not be opened due to an I/O problem.
+     * @throws IOException Thrown, if the output format could not be opened due to an I/O problem.
      */
     @Override
     public void open(int taskNumber, int numTasks) throws IOException {
+        throwable = new AtomicReference<>();
+        this.semaphore = new Semaphore(maxConcurrentRequests);
         this.session = cluster.connect();
-        this.prepared = session.prepare(insertQuery);
         this.callback =
-                new FutureCallback<ResultSet>() {
+                new FutureCallback<V>() {
```

Comment:
Is the type V relevant anywhere? Could we just use a wildcard?

Reply:
Yes, remember (cf PR description) that `CassandraOutputFormat` is no more the base class for row and tuple based output formats, it is now the base class for all the output formats.
`V` is used as  `Void` in Pojo, and as `ResultSet` for `CassandraColumnarOutputFormatBase` (new base class for tuple and row output formats). 
Regarding coding style of type variables vs wildcard I used the same coding style as in `CassandraSinkBase` (type variable). Generally speaking I prefer using type variables when possible rather than wildcard to avoid capture.